src/zkp/whitelist.rs

```diff
@@ -0,0 +1,470 @@
+                offline_keys.as_c_ptr() as *const secp256k1::secp256k1_sys::PublicKey,
+                n_keys,
+                whitelist_key.as_c_ptr(),
+            )
+        };
+        if ret != 1 {
+            return Err(Error::InvalidWhitelistProof);
+        }
+
+        Ok(())
+    }
+
+    /// Obtains a raw const pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_ptr(&self) -> *const ffi::WhitelistSignature {
+        &self.0
+    }
+
+    /// Obtains a raw mutable pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::WhitelistSignature {
+        &mut self.0
+    }
+}
+
+#[cfg(feature = "std")]
+impl fmt::LowerHex for WhitelistSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for ch in self.serialize().iter() {
+            write!(f, "{:02x}", ch)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(feature = "std")]
+impl fmt::Display for WhitelistSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+#[cfg(feature = "std")]
+impl fmt::Debug for WhitelistSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+#[cfg(feature = "std")]
+impl str::FromStr for WhitelistSignature {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<WhitelistSignature, Error> {
+        let mut buf = vec![0; s.len() / 2];
+        from_hex(s, &mut buf).map_err(|_| Error::InvalidWhitelistSignature)?;
+        WhitelistSignature::from_slice(&buf)
+    }
+}
+
+#[cfg(all(feature = "serde", feature = "std"))]
+impl ::serde::Serialize for WhitelistSignature {
+    fn serialize<S: ::serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        if s.is_human_readable() {
+            s.collect_str(self)
+        } else {
+            s.serialize_bytes(&self.serialize())
+        }
+    }
+}
+
+#[cfg(all(feature = "serde", feature = "std"))]
+impl<'de> ::serde::Deserialize<'de> for WhitelistSignature {
+    fn deserialize<D: ::serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        use serde_util;
+
+        if d.is_human_readable() {
+            d.deserialize_str(serde_util::FromStrVisitor::new("an ASCII hex string"))
+        } else {
+            d.deserialize_bytes(serde_util::BytesVisitor::new(
+                "a bytestring",
+                WhitelistSignature::from_slice,
+            ))
+        }
+    }
+}
+
+impl CPtr for WhitelistSignature {
+    type Target = ffi::WhitelistSignature;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+#[cfg(all(test, feature = "global-context"))]
+mod tests {
+    use super::*;
+    use rand::thread_rng;
+    use SECP256K1;
+
+    fn test_whitelist_proof_roundtrip(n_keys: usize) {
```

Comment:
Would prefer to have helpers defined after their usage.

Reply:
Well, it is a bit of a C89-ism if you ask me :)

The way I think about it usually is: When I look at an entire file / module, I want to see the most important things first (usually public API) or in the case of the `tests` module the tests.

While I am reading through the tests / top-level functions, I may come across a function I've not seen before. Assuming a reasonably descriptive name, I can continue reading (and understanding) the original function and thereby follow along the steps of what it does better without having to context-switch to the details of how that functions achieves its job.

If all of the helper functions are defined before they are used, reading the file top to bottom leaves me with a lot of code that is initially unrelated to each other and only once I get to the piece where it is used, it finally comes together.

It is a personal preference but I find the former to be easier :)

In this case however, it is debatable as literally the entire test is a delegation to another function so it might as well be defined at the top :D